include/anvill/Arch.h

```diff
@@ -0,0 +1,233 @@
+          VariantConstraint("AX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("EAX", kTypeIntegral, kMaxBit32),
+          VariantConstraint("RAX", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({
+          VariantConstraint("DL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("DX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("EDX", kTypeIntegral, kMaxBit32),
+          VariantConstraint("RDX", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({
+          VariantConstraint("CL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("CX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("ECX", kTypeIntegral, kMaxBit32),
+          VariantConstraint("RCX", kTypeIntegral, kMaxBit64),
+      }),
+      RegisterConstraint({VariantConstraint("XMM0", kTypeFloat, kMaxBit128)}),
+      RegisterConstraint({VariantConstraint("XMM1", kTypeFloat, kMaxBit128)}),
+
+      // Since the FPU registers are 80 bits wide, they are only able to hold
+      // 64-bit values.
+      RegisterConstraint({VariantConstraint("ST0", kTypeFloat, kMaxBit80)}),
+      RegisterConstraint({VariantConstraint("ST1", kTypeFloat, kMaxBit80)}),
+  };
+};
+
+// This is the cdecl calling convention referenced by llvm::CallingConv::C
+class X86_C : public CallingConvention {
+ public:
+  X86_C(const remill::Arch *arch) : CallingConvention(llvm::CallingConv::C, arch) {}
+  virtual ~X86_C() = default;
+  void AllocateSignature(FunctionDecl &fdecl, const llvm::Function &func);
+  std::vector<ParameterDecl> BindParameters(const llvm::Function &function, bool injected_sret);
+  std::vector<ValueDecl> BindReturnValues(const llvm::Function &function, bool &injected_sret);
+  void BindReturnStackPointer(FunctionDecl &fdecl, const llvm::Function &func);
+
+ private:
+  // Register allocations for parameters are not allowed in vanilla cdecl
+  const std::vector<RegisterConstraint> parameter_register_constraints = {};
+
+  // For x86_C (cdecl), structs can be split over EAX, EDX, ECX, ST0, ST1.
+  const std::vector<RegisterConstraint> return_register_constraints = {
+      RegisterConstraint({
+          VariantConstraint("AL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("AX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("EAX", kTypeIntegral, kMaxBit32),
+      }),
+      RegisterConstraint({
+          VariantConstraint("DL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("DX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("EDX", kTypeIntegral, kMaxBit32),
+      }),
+      RegisterConstraint({
+          VariantConstraint("CL", kTypeIntegral, kMaxBit8),
+          VariantConstraint("CX", kTypeIntegral, kMaxBit16),
+          VariantConstraint("ECX", kTypeIntegral, kMaxBit32),
+      }),
+      RegisterConstraint({VariantConstraint("ST0", kTypeFloat, kMaxBit80)}),
+      RegisterConstraint({VariantConstraint("ST1", kTypeFloat, kMaxBit80)}),
+  };
```

Comment:
Here's some confusing things, probably deserving of some comments in the code:
Vectors (specified in C) of integers/floats can be passed over mmx registers in 32-bit with the gcc, e.g. https://gcc.godbolt.org/z/X9JjUV (gcc), though Clang will use memory (https://gcc.godbolt.org/z/8QX3eW) instead. This seems to be because Clang do ABI lowering of an `<i32, i32>` or `<float, float>` to either an `i64` or a `double`, depending on 32-bit or 64-bit. If we try to force Clang's hand, by manually using vector valued parameters, then it uses XMM registers: https://gcc.godbolt.org/z/C9Ppfg and https://gcc.godbolt.org/z/PH5Q__ for parameter passing.

Reply:
We'll probably want to handle the latter case, of being able to put floats/integer vector values into xmm regs.

Reply:
Maybe someone will respond: https://twitter.com/peter_a_goodman/status/1217942699594207232

Reply:
So, if we decided on the doing the latter case emulating clang, then this was fixed in 
75dfa9a by the same change that lets vectors be passed in xmm regs (see your comment above). I can put a comment if you think its necessary, but if we are emulating clang it seems logical. If clang lowers it for us then we will not try to allocate in xmm